sdcm/cluster_aws.py

```diff
@@ -339,25 +341,33 @@ def update_bootstrap(ec2_user_data, enable_auto_bootstrap):
                 ec2_user_data += ' --bootstrap false '
         return ec2_user_data
 
-    def _create_or_find_instances(self, count, ec2_user_data, dc_idx, az_idx=0):
-        if self.nodes:
-            return self._create_instances(count, ec2_user_data, dc_idx, az_idx)
+    # pylint: disable=too-many-arguments
+    def _create_or_find_instances(self, count, ec2_user_data, dc_idx, az_idx, running=False):
         if self.test_config.REUSE_CLUSTER:
             instances = self._get_instances(dc_idx)
             assert len(instances) == count, f"Found {len(instances)} instances, while expect {count}"
             self.log.info('Found instances to be reused from test [%s] = %s', self.test_config.REUSE_CLUSTER, instances)
             return instances
-        if instances := self._get_instances(dc_idx):
-            self.log.info('Found provisioned instances = %s', instances)
-            return instances
-        self.log.info('Found no provisioned instances. Provision them.')
-        return self._create_instances(count, ec2_user_data, dc_idx, az_idx)
+
+        # pylint: disable=protected-access
+        already_in_use = [node._instance.instance_id for node in self.nodes if node.dc_idx == dc_idx]
+        unused_on_cloud = [inst for inst in self._get_instances(dc_idx, running=running)
+                           if inst.instance_id not in already_in_use]
+        if len(unused_on_cloud) >= count:
+            self.log.info('Found provisioned instances = %s', unused_on_cloud[:count])
+            return unused_on_cloud[:count]
+
+        newly_provisioned = self._create_instances(count - len(unused_on_cloud), ec2_user_data, dc_idx, az_idx=az_idx)
+        self.log.info('Found provisioned instances = %s, and provisioned additional instances = %s ', unused_on_cloud,
+                      newly_provisioned)
+        return unused_on_cloud + newly_provisioned
 
     # pylint: disable=too-many-arguments
     def add_nodes(self, count, ec2_user_data='', dc_idx=0, rack=0, enable_auto_bootstrap=False):
         ec2_user_data = self.prepare_user_data(enable_auto_bootstrap=enable_auto_bootstrap)
 
-        instances = self._create_or_find_instances(count=count, ec2_user_data=ec2_user_data, dc_idx=dc_idx, az_idx=rack)
+        instances = self._create_or_find_instances(count=count, ec2_user_data=ec2_user_data, dc_idx=dc_idx,
```

Comment:
this is the only call to `_create_or_find_instances`, so why do we need the `running` argument ?

Reply:
By default `running=False` and all found instances will be chosen, even the instance that was decommissioned and can not be used (actually, creation of the node on this instance fails). So I filter such instances out

Reply:
again since it's the only call, why do we need a parameter ? is there a use case we want to get the instances that are decommissioned ?

Reply:
Not for existent case. Just to do that more flexible.

Reply:
I would suggest to remove it, i.e. it's more places to reason about it and explain this parameter, and it's not something easy to assume on your own.